executor/insert_common.go

```diff
@@ -274,37 +274,53 @@ func (e *InsertValues) fillDefaultValues(row types.DatumRow, hasValue []bool) er
 	return nil
 }
 
-func (e *InsertValues) getRowsSelectChunk(ctx context.Context, cols []*table.Column) ([]types.DatumRow, error) {
+func (e *InsertValues) getRowsSelectChunk(ctx context.Context, cols []*table.Column, exec func(rows []types.DatumRow) error) error {
 	// process `insert|replace into ... select ... from ...`
 	selectExec := e.children[0]
 	if selectExec.Schema().Len() != len(cols) {
-		return nil, ErrWrongValueCountOnRow.GenByArgs(1)
+		return ErrWrongValueCountOnRow.GenByArgs(1)
 	}
-	var rows []types.DatumRow
 	fields := selectExec.retTypes()
+	maxChunkSize := e.ctx.GetSessionVars().MaxChunkSize
 	for {
 		chk := selectExec.newChunk()
 		iter := chunk.NewIterator4Chunk(chk)
```

Comment:
we can put the `chk` and `iter` outside this `for` loop, and reuse it in the execution of `selectExec`